### PR TITLE
Add a file.link state for hard links

### PR DESCRIPTION
**Pull request by aneeshusa**
### What does this PR do?
Adds a file.link state that creates hard links.

TODO:
- Add some tests
- Add the versionadded tag

### Previous Behavior
No file.link state was available, although there was an execution function.

### New Behavior
There is a file.link state.

### Tests written?
Not yet

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

File: salt/modules/file.py [hunk 1]
```diff
@@ -3179,8 +3179,9 @@ def link(src, path):
     try:
         os.link(src, path)
         return True
-    except (OSError, IOError):
-        raise CommandExecutionError('Could not create \'{0}\''.format(path))
+    except (OSError, IOError) as e:
```
Comment (rallytime) [inline salt/modules/file.py]: I know you said you still have some work to do here, but right off the bat, can you change these `e` references to something like `err` or `error`? We like to have 3-letter variables. See [here](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#variables) for more information.

**Comment by rallytime** [comment]
@aneeshusa There are also some lint errors here, which is causing the tests to fail as well. The name of the broken import is `salt.utils.files`, instead of `file`.

File: salt/states/file.py [hunk 2]
```diff
@@ -1194,6 +1196,207 @@ def _shortcut_check(name,
...
+        'comment': '',
+        'changes': {},
+    }
+
+    if salt.utils.is_windows():
+        return _error(ret, 'The file.link state is not supported on Windows')
+
+    name = os.path.expanduser(name)
+    target = os.path.expanduser(target)
+    if not os.lexists(target):
```
Comment (terminalmage) [inline salt/states/file.py]: This function does not exist, I think you meant `os.path.lexists(target)`

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by aneeshusa** [comment]
Hi, I'm back at Lyft but we don't have a need for this any more, so going to close this now.

**Pull request closed without merging**